test/unittest/dds/participant/ParticipantTests.cpp

```diff
@@ -2170,6 +2171,55 @@ TEST(ParticipantTests, SetListener)
     ASSERT_EQ(DomainParticipantFactory::get_instance()->delete_participant(participant), ReturnCode_t::RETCODE_OK);
 }
 
+class CustomListener2 : public DomainParticipantListener
+{
+public:
+
+    CustomListener2(
+            std::promise<void>& promise)
+        : promise_(promise)
+    {
+    }
+
+    void on_participant_discovery(
+            eprosima::fastdds::dds::DomainParticipant*,
+            eprosima::fastrtps::rtps::ParticipantDiscoveryInfo&&) override
+    {
+        promise_.set_value();
+        std::this_thread::sleep_for(std::chrono::seconds(2));
+    }
+
+private:
+
+    std::promise<void>& promise_;
```

Comment:
I would have not used an external promise but kept it as a member and return the future via a getter. Nevertheless is a style thing and is up to you.